File: jvm/src/main/java/edu/kit/kastel/vads/compiler/parser/visitor/NoOpVisitor.java
package edu.kit.kastel.vads.compiler.parser.visitor;

import de.mr_pine.c0ne.parser.ast.AssignmentTree;
import de.mr_pine.c0ne.parser.ast.BinaryOperationTree;
import de.mr_pine.c0ne.parser.ast.BlockTree;
import de.mr_pine.c0ne.parser.ast.DeclarationTree;
import de.mr_pine.c0ne.parser.ast.FunctionTree;
import de.mr_pine.c0ne.parser.ast.IdentExpressionTree;
import de.mr_pine.c0ne.parser.ast.LValueIdentTree;
import de.mr_pine.c0ne.parser.ast.LiteralTree;
import de.mr_pine.c0ne.parser.ast.NameTree;
import de.mr_pine.c0ne.parser.ast.UnaryOperationTree;
import de.mr_pine.c0ne.parser.ast.ProgramTree;
import de.mr_pine.c0ne.parser.ast.ReturnTree;
import de.mr_pine.c0ne.parser.ast.TypeTree;

/// A visitor that does nothing and returns [Unit#INSTANCE] by default.
/// This can be used to implement operations only for specific tree types.
public interface NoOpVisitor<T> extends Visitor<T, Unit> {

    @Override
    default Unit visit(AssignmentTree assignmentTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(BinaryOperationTree binaryOperationTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(BlockTree blockTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(DeclarationTree declarationTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(FunctionTree functionTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(IdentExpressionTree identExpressionTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(LiteralTree.LiteralIntTree literalTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(LValueIdentTree lValueIdentTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(NameTree nameTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(UnaryOperationTree unaryOperationTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(ProgramTree programTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(ReturnTree returnTree, T data) {
        return Unit.INSTANCE;
    }

    @Override
    default Unit visit(TypeTree typeTree, T data) {
        return Unit.INSTANCE;
    }
}

File: jvm/src/main/java/de/mr_pine/c0ne/parser/visitor/NoOpVisitor.kt
package de.mr_pine.c0ne.parser.visitor

import de.mr_pine.c0ne.parser.ast.*
import de.mr_pine.c0ne.parser.ast.LiteralTree.LiteralBoolTree
import de.mr_pine.c0ne.parser.ast.LiteralTree.LiteralIntTree

/** A visitor that does nothing and returns [Unit] by default.
 * This can be used to implement operations only for specific tree types. */
interface NoOpVisitor<T> : Visitor<T, Unit> {
    override fun visit(assignmentTree: AssignmentTree, data: T) {
    }

    override fun visit(binaryOperationTree: BinaryOperationTree, data: T) {
    }

    override fun visit(blockTree: BlockTree, data: T) {
    }

    override fun visit(declarationTree: DeclarationTree, data: T) {
    }

    override fun visit(functionTree: DeclaredFunctionTree, data: T) {
    }

    override fun visit(identExpressionTree: IdentExpressionTree, data: T) {
    }

    override fun visit(ternaryOperationTree: TernaryOperationTree, data: T) {
    }

    override fun visit(literalIntTree: LiteralIntTree, data: T) {
    }

    override fun visit(literalBoolTree: LiteralBoolTree, data: T) {
    }

    override fun visit(lValueIdentTree: LValueIdentTree, data: T) {
    }

    override fun visit(nameTree: NameTree, data: T) {
    }

    override fun visit(unaryOperationTree: UnaryOperationTree, data: T) {
    }

    override fun visit(programTree: ProgramTree, data: T) {
    }

    override fun visit(ifTree: IfTree, data: T) {
    }

    override fun visit(whileTree: WhileTree, data: T) {
    }

    override fun visit(forTree: ForTree, data: T) {
    }

    override fun visit(breakTree: BreakTree, data: T) {
    }

    override fun visit(continueTree: ContinueTree, data: T) {
    }

    override fun visit(returnTree: ReturnTree, data: T) {
    }

    override fun visit(typeTree: TypeTree, data: T) {
    }

    override fun visit(callTree: CallTree, data: T) {
    }

    override fun visit(builtinFunction: FunctionTree.BuiltinFunction, data: T) {
    }

    override fun <V : Tree> visit(
        parenthesizedListTree: ParenthesizedListTree<V>,
        data: T
    ) {
    }
}
